File: algorithm/line_search/newton.py
import numpy as np

def newton(f, grad_f, hess_f, x0, alpha=1e-4, tol=1e-6, max_iter=10000):
    x = x0.copy()
    history = [x.copy()]

    for k in range(max_iter):
        g = grad_f(x)
        if np.linalg.norm(g) < tol:
            break
        hess = hess_f(x)
        x -= alpha * np.linalg.solve(hess, g)
        history.append(x.copy())

    return x, np.array(history)

File: algorithm/line_search/newton.cpp
#include "newton.h"
#include "../../test_functions/test_function_base.h"
#include <cmath>
#include <numeric>

namespace optimizer {

Newton::Newton(double alpha, double tol, int maxIter)
    : alpha_(alpha), tol_(tol), maxIter_(maxIter)
{}

void Newton::optimize(
    const Eigen::VectorXd& x0,
    const test_function::TestFunction& f
) {
    Eigen::VectorXd x { std::move(x0) };
    history_.push_back(x);

    for (int iter = 0; iter < maxIter_; iter++) {
        Eigen::VectorXd g { f.gradient(x) };
        if (g.norm() < tol_) break;

        Eigen::MatrixXd hess { f.hessian(x) };
        x -= alpha_ * hess.ldlt().solve(g);
        history_.push_back(x);
    }

    result_ = x;
}

}